spec.bs

```diff
@@ -1604,7 +1604,68 @@ For now, see [https://github.com/wicg/observable#operators](https://github.com/w
   The <dfn for=Observable method><code>reduce(|reducer|, |initialValue|, |options|)</code></dfn>
   method steps are:
 
-    1. <span class=XXX>TODO: Spec this and use |reducer|, |initialValue|, and |options|.</span>
+    1. Let |p| [=a new promise=].
+
+    1. Let |controller| be a [=new=] {{AbortController}}.
+
+    1. Let |internal options| be a new {{SubscribeOptions}} whose {{SubscribeOptions/signal}} is the
+       result of [=creating a dependent abort signal=] from the list
+       «|controller|'s [=AbortController/signal=], |options|'s
+       {{SubscribeOptions/signal}} if non-null», using {{AbortSignal}}, and the [=current realm=].
+
+    1. If |internal options|'s {{SubscribeOptions/signal}} is [=AbortSignal/aborted=], then:
+
+       1. [=Reject=] |p| with |internal options|'s {{SubscribeOptions/signal}}'s
+          [=AbortSignal/abort reason=].
+
+       1. Return |p|.
+
+    1. [=AbortSignal/add|Add the following abort algorithm=] to |internal options|'s
+       {{SubscribeOptions/signal}}:
+
+       1. [=Reject=] |p| with |internal options|'s {{SubscribeOptions/signal}}'s [=AbortSignal/abort
+          reason=].
+
+    1. Let |idx| be an {{unsigned long long}}, initially 0.
+
+    1. Let |accumulator| be |initialValue| if it is given, and the string "<code>unset</code>"
```

Comment:
Is this how the Array reduce internals are spec'ed? It seems mildly inefficient.

Most implementations I've seen will either use a `hasValue` boolean, or check the current `index` and whether or not an `initialValue` was provided to determine whether or not to call the reducer.

Reply:
Checking `hasValue` each time a new value is emitted from the Observable will definitely be as inefficient as checking if accumulator is "unset" or some other special internal value each time, since the check is being performed _each time_ regardless, right? So I'm not sure I see an efficiency difference here. But regarding the comment in https://github.com/WICG/observable/pull/171#pullrequestreview-2264670568, I will be migrating to something like `hasValue` anyways.

Reply:
Also FWIW, specs are not concerned with efficiency as much as implementations. They are concerned with whatever semantics are observable to script, and engines are free to optimize things as much as they want or can, provided they have the same observable semantics as specifications.